src/databricks/labs/ucx/migration/azure_credentials.py

```diff
@@ -0,0 +1,249 @@
+                    except UnicodeDecodeError as e:
+                        logger.info(f"Secret {azure_sp_info.secret_scope}.{azure_sp_info.secret_key} does not exists. "
+                                    f"Cannot fetch the service principal client_secret for {azure_sp_info.application_id}. "
+                                    f"Will not reuse this client_secret")
+                except Unauthenticated:
+                    logger.info(f"User is unauthenticated to fetch secret value. Cannot fetch the service principal "
+                                 f"client_secret for {azure_sp_info.application_id}. Will not reuse this client_secret")
+                except PermissionDenied:
+                    logger.info(f"User does not have permission to read secret value for {azure_sp_info.secret_scope}.{azure_sp_info.secret_key}. "
+                                 f"Cannot fetch the service principal client_secret for {azure_sp_info.application_id}. "
+                                 f"Will not reuse this client_secret")
+                except NotFound:
+                    logger.info(f"Secret {azure_sp_info.secret_scope}.{azure_sp_info.secret_key} does not exists. "
+                                 f"Cannot fetch the service principal client_secret for {azure_sp_info.application_id}. "
+                                 f"Will not reuse this client_secret")
+                except InternalError:
+                    logger.info(f"InternalError while reading secret {azure_sp_info.secret_scope}.{azure_sp_info.secret_key}. "
+                                f"Cannot fetch the service principal client_secret for {azure_sp_info.application_id}. "
+                                f"Will not reuse this client_secret")
+
+        # update the list of ServicePrincipalMigrationInfo if client_secret is found
+        for sp in sp_list:
+            if sp.if_managed_identity is False and sp.client_id in azure_sp_info_with_client_secret:
+                sp.client_secret = azure_sp_info_with_client_secret[sp.client_id]
+
+        return
+
+
+    def _save_action_plan(self, sp_list):
+        # save action plan to a file for customer to review, maybe csv format
+        # If SP with replace_with_ac = true, create access connector with same role assignment as the SP, and create SC with the connector
+        # If SP with replace_with_ac = false and client_secret not empty, reuse the client_secret to create SC
+        # If SP with replace_with_ac = false and empty client_secret, create a new client_secret and use it to create SC
+        return
+
+
+    def generate_migration_list(self):
+        """
+        Create the list of SP that need to be migrated, output an action plan as a csv file for users to confirm
+        :return:
+        """
+        # load sp list from azure_storage_account_info.csv
+        loaded_sp_list = self._load_sp_csv()
```

Comment:
Are you saying you were storing a customer secret in a CSV file in plain text? Are you serious?

Reply:
Of course not. client_secret will be took out when save the action plan for user to confirmation. The plan is to use a new dataclass that has less info, but more human readable than `ServicePrincipalMigrationInfo` for action plan.